packages/labs/prefetch/src/client.ts

```diff
@@ -0,0 +1,96 @@
+import throttles from 'throttles';
+import requestIdleCallback from './requestIdleCallback.js';
+
+const events = ['mouseenter', 'touchstart', 'focus'];
+
+const preloaded = new Set<string>();
+
+function shouldPreload({ href }: { href: string }) {
+	try {
+		const url = new URL(href);
+		return (
+			window.location.origin === url.origin &&
+			window.location.pathname !== url.hash &&
+			!preloaded.has(href)
+		);
+	} catch {}
+
+	return false;
+}
+
+let parser: DOMParser;
+let observer: IntersectionObserver;
+
+function observe(link: HTMLAnchorElement) {
+	preloaded.add(link.href);
+	observer.observe(link);
+	events.map((event) => link.addEventListener(event, onLinkEvent, { once: true }));
+}
+
+function unobserve(link: HTMLAnchorElement) {
+	observer.unobserve(link);
+	events.map((event) => link.removeEventListener(event, onLinkEvent));
+}
+
+function onLinkEvent({ target }: Event) {
+	if (!(target instanceof HTMLAnchorElement)) {
+		return;
+	}
+
+	preloadHref(target);
+}
+
+async function preloadHref(link: HTMLAnchorElement) {
```

Comment:
I purposely skipped that check since Firefox and Safari don't support it, but I guess it wouldn't hurt to have it behind a check to avoid prefetching on Chrome for slow connections or data saver mode 👍